### PR TITLE
Validation on single choice tasks

### PR DESCRIPTION
**Pull request by VyrodovMikhail**
Added script for validation of HuggingFace models on single choice tasks. Also, added one test that checks proper prompt generation with k-shot validation.

### CODE REVIEW

File: scripts/validation/validate.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    tokenizer: AutoTokenizer,
+    val_df: pd.DataFrame,
+    device: str,
+    starting_prompt: str,
+    subject_dict: dict[str, str],
+    max_seq_len: int,
+) -> tuple[np.floating[Any], dict[int, int]]:
+    scores = []
+    answers_dict = {}
+    for i in range(len(val_df)):
```
Comment (zemerov) [inline scripts/validation/validate.py:48]: Add progress bar e.g. tqdm

File: requirements.txt [hunk 2]
```diff
@@ -1,5 +1,7 @@
 beautifulsoup4==4.12.2
 numpy==1.26.2
-pandas==2.1.3
+pandas==2.2.1
 requests==2.31.0
 Scrapy==2.11.0
+torch==2.1.1
```
Comment (SpirinEgor) [inline requirements.txt:6]: There is already version 2.2, perhaps it would be a little bit faster

File: scripts/validation/arg_parser.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    parser.add_argument("output_path", type=str, help="Path to the file in which output will be written")
+    parser.add_argument(
+        "-d", "--device", required=False, type=str, help="Can be cpu, some cuda devices or auto", default="cuda"
+    )
+    parser.add_argument(
+        "-prompt",
+        "--prompt_type",
+        required=False,
+        type=str,
+        choices=["mmlu", "mmlu_translated", "not_mmlu"],
```
Comment (SpirinEgor) [inline scripts/validation/arg_parser.py:17]: Maybe custom is a better name

File: scripts/validation/arg_parser.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        default="mmlu",
+    )
+    parser.add_argument(
+        "-dtype",
+        "--torch_dtype",
+        required=False,
+        type=str,
+        choices=["float32", "float16"],
+        help="Type of model's weights",
+    )
```
Comment (SpirinEgor) [inline scripts/validation/arg_parser.py:28]: I don't think we need this, for validation you can always use `float16`

File: scripts/validation/prompts.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"mmlu": "The following are multiple choice questions (with answers) about {}.\n\n", "mmlu_translated": "Ниже приведены вопросы на выбор одного варианта ответа (c ответами) по теме {}.\n\n", "not_mmlu": "Ниже представлены задания на выбор одного варианта ответа (с ответами) по теме {}. Выбери один правильный вариант ответа на последнее задание и напиши его номер.\n\n"}
```
Comment (SpirinEgor) [inline scripts/validation/prompts.json:1]: Could you save it with idents, for now it is hard too read

File: scripts/validation/validate.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    possible_answers = ["1", "2", "3", "4"]
+    tokenized_answers = [tokenizer.convert_tokens_to_ids(str_ans) for str_ans in possible_answers]
+    needed_logits = [logits[0][ans] for ans in tokenized_answers]
+    max_val = max(needed_logits)
+    answer = -1
+    for i in range(4):
+        if needed_logits[i] == max_val:
+            answer = i + 1
+            break
+    return answer
```
Comment (SpirinEgor) [inline scripts/validation/validate.py:34]: Could you just use argmax?

File: scripts/validation/validate.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    input_ids = tokenizer.apply_chat_template(messages, tokenize=True, return_tensors="pt")
+    input_ids = input_ids.to(device)
+    with torch.inference_mode():
+        logits = model(input_ids).logits[:, -1, :]
+    return logits
+
+
+def get_answer(model: AutoModelForCausalLM, tokenizer: AutoTokenizer, task: str, device: str) -> int:
+    logits = get_logits(model, tokenizer, task, device)
+    possible_answers = ["1", "2", "3", "4"]
```
Comment (SpirinEgor) [inline scripts/validation/validate.py:25]: Probably, we need to parametrize this. For example, use letters

File: scripts/validation/validate.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    scores = []
+    answers_dict = {}
+    for i in range(len(val_df)):
+        k_shot_df = get_k_shot_df(val_df, i, args.k_shot)
+        prompt = gen_full_prompt(k_shot_df, val_df, i, starting_prompt, subject_dict, max_seq_len)
+        model_answer = get_answer(model, tokenizer, prompt, device)
+
+        if model_answer not in answers_dict:
+            answers_dict[model_answer] = 0
+        answers_dict[model_answer] += 1
```
Comment (SpirinEgor) [inline scripts/validation/validate.py:55]: Use Counter or defaultdict at least

File: scripts/validation/validate.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+        if model_answer == val_df["answer"][i]:
+            scores.append(1)
+        else:
+            scores.append(0)
```
Comment (SpirinEgor) [inline scripts/validation/validate.py:60]:
```suggestion
        scores.append(model_answer == val_df["answer"][i])
```

File: tests/data/expected_prompt.txt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+The following are multiple choice questions (with answers) about sociology.
+
+В результате хозяйственной деятельности человека увеличилось поступление в атмосферу вредных веществ. Все это негативно влияет на состояние природы и здоровье людей. Иллюстрацией каких глобальных проблем является данный факт?  1) демографических 2) военных 3) экономических 4) экологических
```
Comment (SpirinEgor) [inline tests/data/expected_prompt.txt:3]:
```suggestion
В результате хозяйственной деятельности человека увеличилось поступление в атмосферу вредных веществ. Все это негативно влияет на состояние природы и здоровье людей. Иллюстрацией каких глобальных проблем является данный факт?
1) демографических
2) военных
3) экономических
4) экологических
```

Comment (SpirinEgor) [reply tests/data/expected_prompt.txt:3]: And similar in next shots

File: tests/test_model.py [hunk 11]
```diff
@@ -1,2 +1,2 @@
-def test_dummy():
+def test_dummy() -> None:
     assert 2 + 2 == 4, "Wow, this really shouldn't have happened."
```
Comment (SpirinEgor) [inline tests/test_model.py:2]: You can remove this, it was a placeholder for pytest

**Review by SpirinEgor — CHANGES_REQUESTED** [review]

**Pull request closed without merging**